pykern/fconf.py

```diff
@@ -0,0 +1,569 @@
+begin with a Python word, which can be any Python function. For example::
+
+    count_down:
+      - reversed(range(4))
+
+Produces the following Python Structure::
+
+    {'count_down': [3, 2, 1, 0]}
+
+This allows for a lot of possibilities within the YAML alone.
+
+The Python and YAML macros can call each other. Inside a Python file,
+just use self to call something::
+
+    def count_up(self, top):
+        return reversed(self.count_down(top))
+
+Note that the above count_down example used a special feature of
+result merging. When a Macro returns a list (generator, tuple, etc.)
+in a list context, it will merge with the list. Strings are used in
+place. For example::
+
+    mixed:
+      - range(3)
+      - make_uri('sirepo.com')
+
+Produces::
+
+    {'mixed': [0, 1, 2, 'https://sirepo.com']}
+
+The same is true of dicts. For example::
+
+    fconf_macros:
+      base(num):
+        v${num}.radia.run:
+          description: VM ${num}
+
+    all_hosts:
+      base(3):
+      base(5):
+
+Produces::
+
+    {
+        "all_hosts": {
+            "v3.radia.run": {
+                "description": "VM 3",
+            },
+            "v5.radia.run": {
+                "description": "VM 5",
+            },
+        },
+    }
+
+There are some other useful features. You can refer to any previous
+declared value using variable expansion. For example::
+
+    a:
+      b:
+        c: 3
+    d: ${a}
+    e: ${a.b.c}
+
+
+Produces::
+
+    {
+        'a': {'b': {'c': 3}},
+        'd': {'b': {'c': 3}},
+        'e': 3,
+    }
+
+This allows for flexible constants. Those global strings can be used
+in YAML Macros.
+
+YAML files are evaluated before they are merged. However, they are all
+merged before the next file is evaluated. This allows a main
+"constants" file, for example, to direct the flow of the subsequent files.
+
+:copyright: Copyright (c) 2022 RadiaSoft LLC.  All Rights Reserved.
+:license: http://www.apache.org/licenses/LICENSE-2.0.html
+
+"""
+from pykern.pkcollections import PKDict
+from pykern.pkdebug import pkdc, pkdlog, pkdp
+import contextlib
+import copy
+import inspect
+import pykern.pkrunpy
+import re
+import pykern.pkyaml
+import pykern.pkcollections
+import pykern.pkio
+
+
+#: matches word(word1, ...). Words have to begin with a lettera
+_TEXT_MACRO_DEF = re.compile(
+    r"^([a-z]\w*)\(((?:\s*[a-z]\w*\s*)?(?:,\s*[a-z]\w*\s*)*)\)$", flags=re.IGNORECASE
+)
+
+_MACRO_CALL = re.compile(r"^([a-z]\w*)\((.*)\)$", flags=re.IGNORECASE + re.DOTALL)
+
+_ARG_SEP = re.compile(r"\s*,\s*|\s+")
+
+#: Matches result value and macro argument expansions. FVAR stands for Fconf Variable
+_FVAR = re.compile(r"\$\{(\S+)\}")
+
+#: Use an exact pattern so we can know that the result could be a data structure
+_FVAR_EXACT = re.compile(f"^{_FVAR.pattern}$")
+
+_RESERVED_PREFIX = "fconf_"
+
+_SELF = _RESERVED_PREFIX + "self"
+
+_NO_PARAM = object()
+
+_BUILTINS_EXT = "builtins"
+
+
+def parse_all(path):
```

Comment:
I have noticed that if multiple files in the  directory import logging stuff, as in
```python
from pykern.pkdebug import pkdlog, pkdp
```
then the parser fails with a "duplicate macro" error. I haven't seen it fail with other duplicate imports yet. ???

Reply:
7be8cce